### PR TITLE
Refacto image loading and allow cover configuration

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Read below
| Type?             | refacto
| Category?         | FO
| BC breaks?        | yes - if somebody was using cover_image_id property
| Deprecations?     | no
| How to test?      | See below
| Fixed ticket?     | 
| Related PRs       | https://github.com/PrestaShop/autoupgrade/pull/659
| Sponsor company   |

### Description - performance
- This PR removes excess queries in `getProductProperties` regarding loading images. Everything is now resolved and fetched in `ProductLazyArray`.
- Query count drop on 24 products per page - 710 ➡️ 686

### Description - BC break
- Removes `cover_image_id` property that was used to override id_image. 
- You can still alter the cover by using standard presenter hook.

### Description - new behavior
- It adds a new functionality that allows to choose behavior of cover image for products with combinations.
- Right now, the cover is always forced, even if you present a product with combination specified.
- Now, it allows to specify:
  - If the cover image of products with combinations should be forced to the cover. (current behavior)
  - Or if it should select the first image associated to this combination. This could be useful if filtering module is passing some `id_product_attribute`. So you filter by "Black" attribute and you see all found products with black photos. If the merchant wants to still display the cover, he will just move it so it's the first image in the list.

### How to test
- This concerns only products with combinations.
- If you want to try it, use latest [dev branch of faceted search](https://github.com/PrestaShop/ps_facetedsearch) and enable new option in `Shop settings > Products > Use combination image in listings`.
- With the new option disabled
  - Product image in listings is the cover image all the time.
- With the new option enabled
  - On initial load or if no attribute filter is selected, you will see image of default combination.
  - Once you filter by any attribute, it passes the first `id_product_attribute` it found. You will see the image in more specific color, usually the closest combination found.

### Wording to validate
- Type: Switch
- Name: `Use combination image in listings`
- Helptext: `This option allows you to choose which image you want to display in listings for products with combinations. By default, it always uses the cover image of the product. If you enable this option and your filtering module is properly passing required information, it will display the default image of the combination found.`

### Screenshot
Snímek obrazovky 2023-07-24 200815

### CODE REVIEW

**Review by kpodemski — DISMISSED** [review]

**Comment by kpodemski** [comment]
I'm adding Waiting for PM, so that @MatShir can validate this idea here on GitHub and Waiting for wording, so that we can validate communication to users about this option.

**Comment by Hlavtox** [comment]
Ping @MatShir for final wording, it's the only thing left. 👍

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php [hunk 1]
```diff
@@ -74,6 +74,17 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 ),
                 'required' => false,
             ])
+            ->add('use_combination_image_in_listing', SwitchType::class, [
+                'label' => $this->trans(
+                    'Use combination image in listings',
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'This option allows you to choose which image you want to use in listings for products with combinations. By default, it always uses the cover image of the product. If you enable this option and your filtering module is properly passing required information, you can show directly show the image of this combination.',
```
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php]:
```suggestion
                    'This option allows you to choose which image to display in listings for products with combinations. By default, the cover image of the product will be used. If you enable this option and your filtering module is properly passing the required information, then the default image of the combination found will be displayed.',
```

Comment (l-delin) [reply src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php]: Hello 👋 
I think the wording you added to the PR description is better so I copy pasted that one. Just a few things to edit for clarity. Thanks!

**Review by l-delin — CHANGES_REQUESTED** [review]

**Comment by Hlavtox** [comment]
@l-delin Hi Lea, the wording is applied ;-)

**Review by l-delin — APPROVED** [review]
Looking good, thank you :-)

**Comment by MatShir** [comment]
LGTM ✅

**Comment by Hlavtox** [comment]
@MatShir Can you remove the label Waiting for PM? I can't do it unfortunately 🤣

ping @kpodemski for review

**Comment by MatShir** [comment]
Oups my bad 😆

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php [hunk 2]
```diff
@@ -74,6 +74,17 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 ),
                 'required' => false,
             ])
+            ->add('use_combination_image_in_listing', SwitchType::class, [
+                'label' => $this->trans(
+                    'Use combination image in listings',
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'This option allows you to choose which image to display in listings for products with combinations. By default, the cover image of the product will be used. If you enable this option and your filtering module is properly passing the required information, then the default image of the combination found will be displayed.',
```
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php:83]:
```suggestion
                    'This option allows you to choose which image to display in listings for products with combinations. By default, the cover image of the product will be used. If this option is enabled and your filtering module is properly passing the required information, then the default image of the combination found will be displayed.',
```

**Review by l-delin — COMMENTED** [review]
Disregard, I was testing something 🏋️

**Review by LouisAUTHIE — APPROVED** [review]

File: src/Adapter/Product/PageConfiguration.php [hunk 3]
```diff
@@ -53,6 +53,7 @@ public function getConfiguration()
         return [
             'display_quantities' => $this->configuration->getBoolean('PS_DISPLAY_QTIES'),
             'allow_add_variant_to_cart_from_listing' => $this->configuration->getBoolean('PS_ATTRIBUTE_CATEGORY_DISPLAY'),
+            'use_combination_image_in_listing' => $this->configuration->getBoolean('PS_USE_COMBINATION_IMAGE_IN_LISTING'),
```
Comment (eternoendless) [inline src/Adapter/Product/PageConfiguration.php:56]: You should set the default value in the installer fixtures (install-dev/data/xml/configuration.xml)

Comment (eternoendless) [reply src/Adapter/Product/PageConfiguration.php:56]: And add it during the upgrade

File: classes/Product.php [hunk 4]
```diff
@@ -5549,17 +5549,6 @@ public static function getProductProperties($id_lang, $row, Context $context = n
...
-            $cover = static::getCover($row['id_product']);
-            if (isset($cover['id_image'])) {
-                $row['cover_image_id'] = $cover['id_image'];
-            }
-        }
-
-        if (isset($row['cover_image_id'])) {
-            $cache_key .= '-cover' . (int) $row['cover_image_id'];
-        }
-
```
Comment (eternoendless) [inline classes/Product.php:5562]: I'm not sure about the impact of removing this, just make sure that [this error](https://github.com/PrestaShop/PrestaShop/issues/21389) isn't coming back.

File: src/Adapter/Product/PageConfiguration.php (hunk 3 above)
Comment (Hlavtox) [reply src/Adapter/Product/PageConfiguration.php:56]: @eternoendless It's there, check configuration.xml. :-)

File: classes/Product.php (hunk 4 above)
Comment (Hlavtox) [reply classes/Product.php:5562]: It works fine.
If the template wants the cover -> $product.cover
If it wants a combination image -> $product.default_image

**Review by M0rgan01 — APPROVED** [review]

**Review by florine2623 — APPROVED** [review]
Hello @Hlavtox ,

Tests auto OK : https://github.com/florine2623/testing_pr/actions/runs/6613370369

New option is enabled :
Screenshot 2023-10-24 at 11 56 56

Works as expected ✅ 

Tested with the option enabled / disabled.
Tested with products with one / multiple combinations.
Checked in category page / product page.
Made an order, image is correct in FO and BO

Thanks 💪

**Review by boherm — APPROVED** [review]

**Comment by mflasquin** [comment]
thanks @Hlavtox 👍

**Pull request merged**